### PR TITLE
[Feature] add gptoss continue train bf16-fp8 (sft) example [part1 -  mcore]

### PR DESCRIPTION
**Pull request by yiakwy-xpu-ml-framework-team**
# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

Add gptoss 20b training example in Hopper platform:

Env :

- cu 12.8 + torch 2.9.1
- fa2.8.3 , fa3 (latest)
-  TransformerrEngine  [pytorch ](latest, 2.9.0)
- triton 3.4 (to work with triton sorting kernel)
- Base Image : SGLang 0.5.0rc2

Support parallel scheme:

- PP=1, EP=8
- PP=2, EP=4
- TP=8, ETP=1, EP=8, PP=1

ETP must be 1 ,since ETP should not be > 1 when add_bias is True (required by GptOSS) .

## Snapshot

full gptoss 24 layers with TP8-EP8:
截屏2025-11-27 16 14 54

截屏2025-11-27 16 14 59

## Steps to reproduce

- Generate distributed checkpoint:
   ```
   torchrun $DISTRIBUTED_ARGS convert_mcore_bf16_checkpoint_from_hf.py 2>&1 | tee megatron_fwd.log
   ```
  
- start training jobs
   ```
   # slurm args
   bash training_gptoss_20b_120b_h100_bf16_fp8.sh   
   ```

## Changes

- megatron core:
  - experts : output should be applied bias with unpadded tokens_per_expert 

- GptOss Yarn Config:
  - add GptOss yarn config before model construction

## Other Rleated Issue:

- Megatron Bridge:
  - https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/1491 : fix yarn default parameter
  
- TransformerEngine
  - https://github.com/NVIDIA/TransformerEngine/pull/2445 : build customer transformer-engine to work with pytorch 3.8, pytorch 3.9.1, triton 3.4.0 (safe commit ID for triton kernels : 05b2c186c) (Note transformer uses sort kernel  not supported triton 3.5.1)
  
- SGlang Rollout
  - https://github.com/sgl-project/sglang/pull/14099
  
- VeRL
  - https://github.com/volcengine/verl/pull/4394
  - https://github.com/volcengine/verl/pull/4323

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/models/gpt/gpt_model.py [hunk 1]
```diff
@@ -163,6 +163,25 @@ def __init__(
             )
 
         elif self.position_embedding_type == 'yarn':
+            from megatron.training import get_args, print_rank_0
+
+            args = get_args()
```
Comment (ISEEKYAN) [inline megatron/core/models/gpt/gpt_model.py]: args should not exist in GPTModel. Because args is a component of `training` instead of `core`.

File: megatron/core/models/gpt/gpt_model.py [hunk 2]
```diff
@@ -163,6 +163,25 @@ def __init__(
...
+
+            # Handle GPT-OSS mode with YaRN RoPE configuration
+            if hasattr(args, 'enable_gpt_oss') and args.enable_gpt_oss:
+                print_rank_0("GPT-OSS mode enabled: Configuring YaRN RoPE parameters")
+
+                # Set GPT-OSS YaRN values directly on the config
+                # These defaults are based on Huggingface GPT-OSS configurations
+                config.position_embedding_type = "yarn"
```
Comment (ISEEKYAN) [inline megatron/core/models/gpt/gpt_model.py]: we should not modify config inside the GPTModel, instead we should do this before  entering GPTModel.

Comment (yiakwy-xpu-ml-framework-team) [reply megatron/core/models/gpt/gpt_model.py]: Fully agree. Let me update.

File: megatron/core/models/gpt/gpt_model.py (hunk 1 above)
Comment (yiakwy-xpu-ml-framework-team) [reply megatron/core/models/gpt/gpt_model.py]: resolved.

**Comment by sbhavani** [comment]
@yiakwy-xpu-ml-framework-team thanks for creating this example! 

We are currently refactoring config management to improve validation and reduce the number of default args just like in [Megatron Bridge](https://github.com/NVIDIA-NeMo/Megatron-Bridge/). I'd recommend we wait until the refactor is done this month and then merge the example.

I also noticed you require other changes outside of examples/ and arguments.py. I think it'd better to split those changes into separate PRs (experts.py bug fix, doc fixes, etc.) from the GPT-OSS example.

File: examples/gptoss/convert_mcore_bf16_checkpoint_from_hf.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
+from argparse import ArgumentParser
+
+from megatron.bridge import AutoBridge
+from megatron.bridge.utils.common_utils import get_last_rank, print_rank_0
+from megatron.bridge.training.model_load_save import load_megatron_model, save_megatron_model, load_tokenizer
```
Comment (yaoyu-33) [inline examples/gptoss/convert_mcore_bf16_checkpoint_from_hf.py:5]: Hi, it's not ideal to import bridge from megatron-lm side, it will cause a cyclic dependency. Also megatron-lm env doesnt require users to install bridge.

do you think it's okay to put the example in bridge?

File: examples/llama/convert_megatron_bf16_checkpoint_from_hf.sh [hunk 4]
```diff
@@ -0,0 +1,43 @@
+#/usr/bin/bash
```
Comment (yaoyu-33) [inline examples/llama/convert_megatron_bf16_checkpoint_from_hf.sh:1]: llama changes should be another pr, if this pr is mostly gpt-oss related.

File: megatron/core/parallel_state.py [hunk 5]
```diff
@@ -527,7 +527,7 @@ def initialize_model_parallel(
     order: str = "tp-cp-ep-dp-pp",
     get_embedding_ranks: Optional[Callable[[List[int], Optional[int]], List[int]]] = None,
     get_position_embedding_ranks: Optional[Callable[[List[int], Optional[int]], List[int]]] = None,
-    create_gloo_process_groups: bool = True,
+    create_gloo_process_groups: bool = False,
```
Comment (yaoyu-33) [inline megatron/core/parallel_state.py:530]: this is a major change that affect all training, it will increase the mem footprint, what's the reason?

File: examples/gptoss/convert_mcore_bf16_checkpoint_from_hf.py (hunk 3 above)
Comment (yiakwy-xpu-ml-framework-team) [reply examples/gptoss/convert_mcore_bf16_checkpoint_from_hf.py:5]: @yaoyu-33 I see, this is a tool in example indendent from Megatron-Core. I didn't encoutner the cyclic reference problem. 

I use this script to generate megatron distributed checkpoint and verify it.

File: examples/llama/convert_megatron_bf16_checkpoint_from_hf.sh (hunk 4 above)
Comment (yiakwy-xpu-ml-framework-team) [reply examples/llama/convert_megatron_bf16_checkpoint_from_hf.sh:1]: I noticed we only added verification for llama3. I just updated it from my local repo.

The mian focus this script is gptoss. Yes it is better to be in separate PR but if it is not too inconvenient to you and we can add it to the repo in this PR D.

File: megatron/core/parallel_state.py (hunk 5 above)
Comment (yiakwy-xpu-ml-framework-team) [reply megatron/core/parallel_state.py:530]: Ok I will mute it if memory footprint is increased. I didn't observed side effects and I just find that by deault many gloo connection was created in CPU side.

File: examples/gptoss/convert_mcore_bf16_checkpoint_from_hf.py (hunk 3 above)
Comment (yaoyu-33) [reply examples/gptoss/convert_mcore_bf16_checkpoint_from_hf.py:5]: it's more like we are not asking megatron-lm users to install bridge, so it's a bit confusing to put things here. I think it needs other people's opinions to see what's the best way.

**Comment by fengxy-03** [comment]
@yiakwy-xpu-ml-framework-team @yaoyu-33 Hi, I've implemented a GPT-OSS model (0.11B) based on your guidelines using Megatron-LM 0.16.0rc0. During training, the throughput per GPU is only around 1.0 TFLOP/s, which seems abnormal for this configuration.

Could you please take a look at my script and logs to see if there are any obvious misconfigurations?
#### Training Script 
```python
SEQ_LENGTH=8192
MAX_LENGTH=8192
TRAIN_SAMPLES=1518124
LAST_TRAIN_SAMPLES=0
LR_DECAY_SAMPLES=$(((TRAIN_SAMPLES - LAST_TRAIN_SAMPLES) * 80 / 100))
CHECKPOINT_PATH=
# ================end================
TOKENIZER_TYPE=SentencePieceTokenizer
TOKENIZER_MODEL=
MICRO_BATCH_SIZE=1
GLOBAL_BATCH_SIZE=128

DISTRIBUTED_ARGS=" \
    --nnodes=1 \
    --nproc_per_node=8 \
    --node_rank=0 \
    --master_addr=localhost \
    --master_port=6000"

MODEL_ARGS=" \
    --no-masked-softmax-fusion \
    --transformer-impl transformer_engine \
    --disable-bias-linear \
    --untie-embeddings-and-output-weights \
    --no-rope-fusion \
    --normalization RMSNorm \
    --num-layers 12 \
    --hidden-size 512 \
    --ffn-hidden-size 2048 \
    --num-attention-heads 64 \
    --group-query-attention \
    --num-query-groups 8 \
    --seq-length 8192 \
    --max-position-embeddings 8192 \
    --use-mcore-models \
    --rotary-percent 1.0 \
    --rope-type yarn \
    --position-embedding-type yarn
    --rotary-base 10000 \
    --no-bias-gelu-fusion \
    --export-force-local-attention \
    --no-bias-dropout-fusion \
    --quick-geglu \
    --glu-linear-offset 1.0 \
    --softmax-type learnable \
    --window-attn-skip-freq 2 \
    --activation-func-clamp-value 7.0 \
    --window-size 128,0 \
    --enable-gpt-oss"

MOE_ARGS=" \
    --num-experts 4 \
    --moe-router-topk 2 \
    --moe-router-load-balancing-type aux_loss \
    --moe-aux-loss-coeff 1e-3 \
    --moe-grouped-gemm \
    --moe-token-dispatcher-type alltoall \
    --overlap-param-gather \
    --overlap-grad-reduce \
    --moe-ffn-hidden-size 2048 \
    --moe-router-dtype fp32 \
    --moe-z-loss-coeff 1e-3 \
    --moe-permute-fusion"

DATA_ARGS=" \
    --num-workers 8 \
    --dataloader-type cyclic \
    --tokenizer-type ${TOKENIZER_TYPE} \
    --tokenizer-model ${TOKENIZER_MODEL} \
    --data-path \
    --split 1000,0,0 \
    --no-create-attention-mask-in-dataloader"

TRAINING_ARGS=" \
    --micro-batch-size ${MICRO_BATCH_SIZE} \
    --global-batch-size ${GLOBAL_BATCH_SIZE} \
    --lr 1.0e-5 \
    --train-samples ${TRAIN_SAMPLES} \
    --lr-decay-samples ${LR_DECAY_SAMPLES} \
    --lr-decay-style cosine \
    --min-lr 1.0e-6 \
    --weight-decay 0.1 \
    --lr-warmup-fraction 0.05 \
    --clip-grad 1.0 \
    --bf16 \
    --use-flash-attn \
    --attention-softmax-in-fp32 \
    --accumulate-allreduce-grads-in-fp32 \
    --disable-bf16-reduced-precision-matmul \
    --recompute-activations"

MODEL_PARALLEL_ARGS=" \
    --tensor-model-parallel-size 4 \
    --pipeline-model-parallel-size 1 \
    --expert-model-parallel-size 2 \
    --sequence-parallel \
    --context-parallel-size 1 \
    --use-distributed-optimizer \
    --fp8-format hybrid \
    --fp8-param-gather \
    --fp8-amax-compute-algo max \
    --fp8-amax-history-len 1024"
    
LOGGING_ARGS=" \
    --log-interval 1 \
    --save-interval 10000 \
    --eval-interval 50000000 \
    --eval-iters 0 \
    --save $CHECKPOINT_PATH \
    --tensorboard-dir "${CHECKPOINT_PATH}/tensorboard" \
    --wandb-project ${WANDB_PROJECT:-"gpt-oss"} \
    --wandb-exp-name ${WANDB_NAME:-"gpt-oss-test"} \
    --moe-per-layer-logging \
    --no-load-optim \
    --no-load-rng \
    --log-throughput"


python -m torch.distributed.run ${DISTRIBUTED_ARGS} pretrain_gpt.py \
    ${MODEL_ARGS} \
    ${MOE_ARGS} \
    ${DATA_ARGS} \
    ${TRAINING_ARGS} \
    ${MODEL_PARALLEL_ARGS} \
    ${LOGGING_ARGS}
```
#### Logs
```
 [2026-01-16 15:01:05] iteration        3/   11860 | consumed samples:          384 | elapsed time per iteration (ms): 75210.0 | throughput per GPU (TFLOP/s/GPU): 1.1 | learning rate: 6.323595E-08 | global batch size:   128 | lm loss: 6.102232E+00 | z_loss: 2.145113E+00 | load_balancing_loss: 1.155455E+00 | loss scale: 1.0 | grad norm: 75.384 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2026-01-16 15:02:20] iteration        4/   11860 | consumed samples:          512 | elapsed time per iteration (ms): 75126.0 | throughput per GPU (TFLOP/s/GPU): 1.1 | learning rate: 8.431460E-08 | global batch size:   128 | lm loss: 6.109521E+00 | z_loss: 2.139967E+00 | load_balancing_loss: 1.154540E+00 | loss scale: 1.0 | grad norm: 74.716 | number of skipped iterations:   0 | number of nan iterations:   0 |
```

**Comment by Phlip79** [comment]
Please reference [Megatron-Bridge](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/src/megatron/bridge/recipes/gpt_oss/gpt_oss.py) for how to use bf16 and fp8 for GPT-OSS.
